### PR TITLE
feat: make auto-approve handle build.gradle dependency upgrades and apiary versioning schemes

### LINKED ISSUE #2424: auto-approve: please add the following patterns to auto approve
* `com.google.cloud.tools:`     GoogleCloudPlatform/java-docs-samples#5890
* `com.google.cloud.functions:` GoogleCloudPlatform/java-docs-samples#5889
* `com.google.cloud.*:` Would probably be good as well.
* `com.google.apis:`  GoogleCloudPlatform/java-docs-samples#5888

### PR DESCRIPTION
**Pull request by sofisl**
Fixes #2424 and #2371 

cc @lesv

### CODE REVIEW

File: packages/auto-approve/src/utils-for-pr-checking.ts [hunk 1]
```diff
@@ -472,19 +442,54 @@ export function getJavaVersions(
...
+      oldMajorVersion = oldVersions[5] || oldVersions[7];
+      oldMinorVersion = oldVersions[6] || oldVersions[8];
+    }
+
+    if (newVersions) {
+      newDependencyName = `${newVersions[1]}:${newVersions[2]}`;
+      newMajorRevVersion = newVersions[5];
+      newMajorVersion = newVersions[6] || newVersions[8];
+      newMinorVersion = newVersions[7] || newVersions[9];
+    }
```
Comment (lesv) [inline packages/auto-approve/src/utils-for-pr-checking.ts:476]: It would be nice to see a comment about what's going on here.

**Review by lesv — APPROVED** [review]

File: packages/auto-approve/src/interfaces.ts [hunk 2]
```diff
@@ -54,6 +55,16 @@ export interface Versions {
   newMinorVersion: string;
 }
 
+/**
+ * Interface for return type of array from getTargetFiles. It is the amalgamation
+ * of the file that was changed in the PR, plus the file rule that applies
+ * to that given file.
+ */
+export interface fileAndMetadata {
```
Comment (chingor13) [inline packages/auto-approve/src/interfaces.ts:63]:
```suggestion
export interface FileAndMetadata {
```

Let's keep interfaces/classes capitalized

File: packages/auto-approve/src/utils-for-pr-checking.ts [hunk 3]
```diff
@@ -472,19 +442,54 @@ export function getJavaVersions(
...
-  if (oldVersions) {
-    oldDependencyName = `${oldVersions[1]}:${oldVersions[2]}`;
-    oldMajorVersion = oldVersions[4] || oldVersions[6];
-    oldMinorVersion = oldVersions[5] || oldVersions[7];
+
+  if (versionFile.filename.includes('build.gradle')) {
+    if (oldVersions) {
+      oldDependencyName = oldVersions[1] || oldVersions[4];
+      oldMajorVersion = oldVersions[2] || oldVersions[5];
+      oldMinorVersion = oldVersions[6] || oldVersions[3];
```
Comment (chingor13) [inline packages/auto-approve/src/utils-for-pr-checking.ts:455]: It might be nice to use named group captures so this can be read more easily.

**Review by chingor13 — APPROVED** [review]

File: packages/auto-approve/src/process-checks/java/dependency.ts [hunk 4]
```diff
@@ -82,10 +82,10 @@ export class JavaDependency extends Process implements LanguageRule {
                   or
                   <groupId>com.google.apis</groupId>
                   <artifactId>google-api-services-policytroubleshooter</artifactId>
-            -     <version>v1-rev20210319-1.31.5</version>
+                  -     <version>v1-rev20210319-1.31.5</version>
                 */
             oldVersion: new RegExp(
-              /<groupId>([^<]*)<\/groupId>[\s]*<artifactId>([^<]*)<\/artifactId>[\s]*-[\s]*<version>(v[0-9]-rev[0-9]*-([0-9]*)\.([0-9]*\.[0-9])|([0-9]*)\.([0-9]*\.[0-9]*))<\/version>[\s]*/
+              /<groupId>([^<]*)<\/groupId>[\s]*<artifactId>([^<]*)<\/artifactId>[\s]*-[\s]*<version>(v[0-9]-rev([0-9]*)-([0-9]*)\.([0-9]*\.[0-9])|([0-9]*)\.([0-9]*\.[0-9]*))<\/version>[\s]*/
```
[the remark this answers is not in the document]
Comment (sofisl) [reply packages/auto-approve/src/process-checks/java/dependency.ts:88]: I'm not sure I know how to do that! Could you give an example? I wasn't able to combine by just creating variables and concatenating them together.

**Pull request merged**